app/scripts/lib/transaction/handlers.ts

```diff
@@ -0,0 +1,206 @@
+import type { Provider } from '@metamask/network-controller';
+import { FetchGasFeeEstimateOptions } from '@metamask/gas-fee-controller';
+import { safelyExecute } from '@metamask/controller-utils';
+
+import {
+  TokenStandard,
+  TransactionMeta,
+} from '../../../../shared/constants/transaction';
+import {
+  MetaMetricsEventFragment,
+  MetaMetricsPageObject,
+  MetaMetricsReferrerObject,
+} from '../../../../shared/constants/metametrics';
+import { type SnapAndHardwareMessenger } from '../snap-keyring/metrics';
+
+import * as metrics from './metrics';
+
+export type TransactionEventRequest = {
+  createEventFragment: (
+    options: MetaMetricsEventFragment,
+  ) => MetaMetricsEventFragment;
+  finalizeEventFragment: (
+    fragmentId: string,
+    options?: {
+      abandoned?: boolean;
+      page?: MetaMetricsPageObject;
+      referrer?: MetaMetricsReferrerObject;
+    },
+  ) => void;
+  getEventFragmentById: (fragmentId: string) => MetaMetricsEventFragment;
+  updateEventFragment: (
+    fragmentId: string,
+    payload: Partial<MetaMetricsEventFragment>,
+  ) => void;
+  getAccountType: (
+    address: string,
+  ) => Promise<'hardware' | 'imported' | 'MetaMask'>;
+  getDeviceModel: (
+    address: string,
+  ) => Promise<'ledger' | 'lattice' | 'N/A' | string>;
+  // According to the type GasFeeState returned from getEIP1559GasFeeEstimates
+  // doesn't include some properties used in buildEventFragmentProperties,
+  // hence returning any here to avoid type errors.
+  getEIP1559GasFeeEstimates(options?: FetchGasFeeEstimateOptions): Promise<any>;
+  getParticipateInMetrics: () => boolean;
+  getSelectedAddress: () => string;
+  getTokenStandardAndDetails: () => {
+    decimals?: string;
+    balance?: string;
+    symbol?: string;
+    standard?: TokenStandard;
+  };
+  getTransaction: (transactionId: string) => TransactionMeta;
+  provider: Provider;
+  snapAndHardwareMessenger: SnapAndHardwareMessenger;
+  trackEvent: (payload: any) => void;
+};
+
+export type TransactionEventPayload = {
+  transactionMeta: TransactionMeta;
+  actionId?: string;
+  error?: string;
+  approvalTransactionMeta?: TransactionMeta;
+};
+
+/**
+ * This function is called when a transaction is added to the controller.
+ *
+ * @param TransactionEventRequest - Contains controller actions needed to create/update/finalize event fragments
+ * @param transactionEventPayload - The event payload
+ * @param transactionEventPayload.transactionMeta - The transaction meta object
+ */
+export const handleTransactionAdded = async (
```

Comment:
I though this would be a nice idea to separate these handlers on it's own file and scope. Metrics are the only domain that cares about transaction events for now but in future I believe there could be more.